extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,238 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Token } from 'markdown-it';
+import * as vscode from 'vscode';
+import { MarkdownEngine } from '../markdownEngine';
+import { TableOfContentsProvider, TocEntry } from '../tableOfContentsProvider';
+
+export default class MarkdownSmartSelect implements vscode.SelectionRangeProvider {
+
+	constructor(
+		private readonly engine: MarkdownEngine
+	) { }
+
+	public async provideSelectionRanges(document: vscode.TextDocument, positions: vscode.Position[], _token: vscode.CancellationToken): Promise<vscode.SelectionRange[] | undefined> {
+		let promises = await Promise.all(positions.map((position) => {
+			return this.provideSelectionRange(document, position, _token);
+		}));
+		return promises.filter(item => item !== undefined) as vscode.SelectionRange[];
+	}
+
+	private async provideSelectionRange(document: vscode.TextDocument, position: vscode.Position, _token: vscode.CancellationToken): Promise<vscode.SelectionRange | undefined> {
+		const headerRange = await this.getHeaderSelectionRange(document, position);
+		const blockRange = await this.getBlockSelectionRange(document, position, headerRange);
+		return blockRange ? blockRange : headerRange ? headerRange : undefined;
+	}
+
+	private async getBlockSelectionRange(document: vscode.TextDocument, position: vscode.Position, headerRange?: vscode.SelectionRange): Promise<vscode.SelectionRange | undefined> {
+
+		const tokens = await this.engine.parse(document);
+
+		let blockTokens = getTokensForPosition(tokens, position);
+
+		if (blockTokens.length === 0) {
+			return undefined;
+		}
+
+		let parentRange = headerRange ? headerRange : createBlockRange(document, position.line, blockTokens.shift());
+		let currentRange: vscode.SelectionRange | undefined;
```

Comment:
[nit] Could you use a single variable instead of separate current and parent vars? Based on the loop, I suspect that `parentRange === currentRange` for most of the time